src/main/java/trackbot/ui/Parser.java

```diff
@@ -0,0 +1,205 @@
+                throw TrackBotException.invalidFormat("deadline", "deadline <task description> /by <date/time>");
+            }
+            return new AddCommand(new Deadline(deadlineParts[0], deadlineParts[1]));
+        case "event":
+            String[] eventParts = otherWord.split(" /from | /to ");
+            if (eventParts.length < 3) {
+                throw TrackBotException.invalidFormat("event", "event <description> /from <start> /to <end>");
+            }
+            return new AddCommand(new Event(eventParts[0], eventParts[1], eventParts[2]));
+        case "delete":
+            if (otherWord.isEmpty()) {
+                throw TrackBotException.invalidFormat("delete", "delete <task number>");
+            }
+            try {
+                int num = Integer.parseInt(otherWord) - 1;
+                return new DeleteCommand(num);
+            } catch (NumberFormatException e) {
+                throw TrackBotException.invalidFormat("delete", "delete <task number>");
+            }
+        case "mark":
+            if (otherWord.isEmpty()) {
+                throw TrackBotException.invalidFormat("mark", "mark <task number>");
+            } else {
+                try {
+                    int num = Integer.parseInt(otherWord) - 1;
+                    return new MarkCommand(num);
+                } catch (NumberFormatException e) {
+                    throw TrackBotException.invalidFormat("mark", "mark <task number>");
+                }
+            }
+        case "unmark":
+            if (otherWord.isEmpty()) {
+                throw TrackBotException.invalidFormat("unmark", "unmark <task number>");
+            } else {
+                try {
+                    int num = Integer.parseInt(otherWord) - 1;
+                    return new UnmarkCommand(num);
+                } catch (NumberFormatException e) {
+                    throw TrackBotException.invalidFormat("unmark", "unmark <task number>");
+                }
+            }
+        case "find":
+            if (otherWord.isEmpty()) {
+                throw TrackBotException.invalidFormat("find", "find <task keyword>");
+            }
+            return new FindCommand(otherWord);
+        case "bye":
+            return new ExitCommand();
+        default:
+            throw new TrackBotException("Sorry, I did not understand that command.");
+        }
+    }
+
+    /**
+     * Parses a task string from storage into a Task object.
+     *
+     * @param line The raw string representing a task in storage format.
+     * @return The Task object corresponding to the input string, or null if the format is invalid.
+     */
+    public static Task parseTask(String line) {
+        String[] parts = line.split(" \\| ");
+        String taskType = parts[0];
+        boolean isDone = parts[1].equals("1");
+        Task task;
+
+        switch (taskType) {
+        case "T":
+            task = new ToDo(parts[2]);
+            break;
+        case "D":
```

Comment:
Switch code block shouldn't have an extra indentation, refer to previous comment for example.